propolis-server/src/initializer.rs

```diff
@@ -0,0 +1,287 @@
+use anyhow::Result;
+use std::fs::File;
+use std::io::{Error, ErrorKind};
+use std::sync::Arc;
+
+use propolis::bhyve_api;
+use propolis::block;
+use propolis::chardev::{Sink, Source, UDSock};
+use propolis::common::PAGE_SIZE;
+use propolis::dispatch::Dispatcher;
+use propolis::hw::chipset::{i440fx::I440Fx, Chipset};
+use propolis::hw::ibmpc;
+use propolis::hw::pci;
+use propolis::hw::ps2ctrl::PS2Ctrl;
+use propolis::hw::qemu::{debug::QemuDebugPort, fwcfg, ramfb};
+use propolis::hw::uart::LpcUart;
+use propolis::hw::virtio;
+use propolis::instance::Instance;
+use propolis::inventory::{EntityID, Inventory};
+use propolis::vmm::{self, Builder, Machine, MachineCtx, Prot};
+
+use crate::Serial;
+
+// Arbitrary ROM limit for now
+const MAX_ROM_SIZE: usize = 0x20_0000;
+
+fn open_bootrom<P: AsRef<std::path::Path>>(path: P) -> Result<(File, usize)> {
+    let fp = File::open(path.as_ref())?;
+    let len = fp.metadata()?.len();
+    if len % (PAGE_SIZE as u64) != 0 {
+        Err(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "rom {} length {:x} not aligned to {:x}",
+                path.as_ref().to_string_lossy(),
+                len,
+                PAGE_SIZE
+            ),
+        )
+        .into())
+    } else {
+        Ok((fp, len as usize))
+    }
+}
+
+pub fn build_instance(
+    name: &str,
+    max_cpu: u8,
+    lowmem: usize,
+) -> Result<Arc<Instance>> {
+    let builder = Builder::new(name, true)?
+        .max_cpus(max_cpu)?
+        .add_mem_region(0, lowmem, Prot::ALL, "lowmem")?
```

Comment:
The non-answer-but-sorta-truth is https://github.com/oxidecomputer/propolis/blob/master/propolis-cli/src/main.rs#L42-L65

@pfmooney please correct me, but I *think* some of it is arbitrary, and other parts are embedded within the expectations of the machine's layout.

- `vmm::MAX_SYSMEM` and `vmm:MAX_PHYSMEM` are documented as "arbitrary"
- The ROM region starting behind `0x1_0000_0000` is almost certainly related to the `0xFFFF_FFF0` x86 boot start constraint (though I'm not 100% sure why there's that `0x10` byte gap; I guess x86 CPUs actually like to boot into the last 16 bytes of bootrom).
- For the rest (mmio regions) I *think* they're arbitrary. If there are constraints, it would definitely be good to encode them.